### ISSUE TITLE
High CPU and memory consumption with buffer enabled

### ISSUE THREAD

**Issue by alpi-ua** [timberio/vector#7246]
### Vector Version

```
vector 0.13.0 (v0.13.0 x86_64-unknown-linux-gnu 2021-04-21)
```
### Vector Configuration File

```toml
[api]                                                                                                                                                                                                                                                                                       
  enabled = true                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                            
[sources.journal]                                                                                                                                                                                                                                                                           
  type = "journald"                                                                                                                                                                                                                                                                         
  exclude_units = ["kernel","vector","telegraf","chronyd","crond","init.scope","dnf-makecache.service","rsyslog"]                                                                                                                                                                           
                                                                                                                                                                                                                                                                                            
[sources.app_std]                                                                                                                                                                                                                                                                         
  type = "file"                                                                                                                                                                                                                                                                             
  include = ["/opt/app/*/logs/stdout.log"]                                                                                                                                                                                                                                           
  fingerprint.strategy = "device_and_inode"                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                            
  [sources.app_std.multiline]                                                                                                                                                                                                                                                             
    start_pattern = '^[0-9]{4}-[0-9]{2}-[0-9]{2}'                                                                                                                                                                                                                                           
     mode = "halt_before"                                                                                                                                                                                                                                                                   
     condition_pattern = '^[0-9]{4}-[0-9]{2}-[0-9]{2}'                                                                                                                                                                                                                                      
     timeout_ms = 999999999                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                            
[sources.app]                                                                                                                                                                                                                                                                             
  type = "file"                                                                                                                                                                                                                                                                             
  include = ["/opt/app/*/logs/*/*.log"]                                                                                                                                                                                                                                              
  fingerprint.strategy = "device_and_inode"                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                            
  [sources.app.multiline]                                                                                                                                                                                                                                                                 
     start_pattern = '^[0-9]{4}-[0-9]{2}-[0-9]{2}'                                                                                                                                                                                                                                          
     mode = "halt_before"                                                                                                                                                                                                                                                                   
     condition_pattern = '^[0-9]{4}-[0-9]{2}-[0-9]{2}'                                                                                                                                                                                                                                      
     timeout_ms = 999999999                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                            
[transforms.app_filter]                                                                                                                                                                                                                                                                   
  type = "filter"                                                                                                                                                                                                                                                                           
  inputs = ["app"]                                                                                                                                                                                                                                                                        
  condition.type = "check_fields"                                                                                                                                                                                                                                                           
  condition."message.not_contains" = "NOT_FOUND - no exchange"                                                                                                                                                                                                                              
  condition."message.not_regex" = " (INFO|DEBUG|TRACE) "                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                            
[sinks.app_prod]                                                                                                                                                                                                                                                                          
  type = "vector"                                                                                                                                                                                                                                                                           
  inputs = ["app_filter","app_std"]                                                                                                                                                                                                                                                     
  address = "server:5001"                                                                                                                                                                                                                                                    
  healthcheck = true                                                                                                                                                                                                                                                                        
#  buffer.max_size = 10100000                                                                                                                                                                                                                                                               
#  buffer.type = "disk"                                                                                                                                                                                                                                                                     
#  buffer.when_full = "block"                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                            
[sinks.system]                                                                                                                                                                                                                                                                              
  type = "vector"                                                                                                                                                                                                                                                                           
  inputs = ["journal"]                                                                                                                                                                                                                                                                      
  address = "server:5000"                                                                                                                                                                                                                                                    
  healthcheck = true                                                                                                                                                                                                                                                                        
#  buffer.max_size = 10100000                                                                                                                                                                                                                                                               
#  buffer.type = "disk"                                                                                                                                                                                                                                                                     
#  buffer.when_full = "block"
```

### Debug Output

```
https://gist.github.com/alpi-ua/0a128a6868f20595538908a50e8e5871
```
### Expected Behavior

Vector releasing resources while not producing logs;
Vector working properly with buffers

### Actual Behavior
Vector consuming resources while idle when using buffers.
If I'm not using disk buffer - Vector is working fine

### Additional info:
With lsof I can see that Vector is holding old buffers which supposed to be deleted when it have a stable connection with an output.
After purging buffers directory, new files are created.

![image](https://user-images.githubusercontent.com/14225967/116065664-e9eba200-a68f-11eb-8174-8b104760f66f.png)
![image](https://user-images.githubusercontent.com/14225967/116065781-08519d80-a690-11eb-9b26-65adf1421a27.png)

### References
https://github.com/timberio/vector/issues/3594
https://github.com/timberio/vector/discussions/7159

**Comment by ryn9** [comment]
I am seemingly having the same issue with 0.13.1 (v0.13.1 x86_64-unknown-linux-musl 2021-04-29)

**Comment by ryn9** [comment]
Was asked to provide more info.....

Here is a sanitized version of the config I am using..

```
api:
  enabled: true

log_schema:
  timestamp_key: "timestamp"

sources:
  file_input:
    type: file
    include:
    - "/tmp/file_input/*.ndjson*"
    read_from: beginning

transforms:
  remap_transform:
    inputs:
      - file_input
    type: remap
    source: |2-
          . = parse_json!(string!(.message))
          #sanitized
  filter_transform:
    type: filter
    inputs:
      - remap_transform
    condition: .type != null

sinks:
  out:
    type: http
    inputs:
      - filter_transform
    compression: gzip
    uri: https://destination.host.com:12345/
    tls:
      ca_file: /etc/vector/ssl_auth/ca-certificates.crt
      crt_file: /etc/vector/ssl_auth/client.crt
      key_file: /etc/vector/ssl_auth/client.key
      verify_hostname: true
    encoding:
      codec: ndjson
      timestamp_format: rfc3339
    buffer:
      max_size: 10737418240
      type: disk
      when_full: block
    request:
      concurrency: 6
```

Additionally - It is worth mentioning the service was seen using more and more cpu over the period of a few days. Restarting the service did not clear up the cpu utilization. Stopping the service, clearing the data_dir associated with the out_buffer, and then starting the service back up - did finally bring the cpu back to the measured cpu from when the service was first started.

**Comment by jszwedko** [comment]
I was hoping to reproduce this more quickly with high throughput test, but I haven't been able to just yet. I'll try to set something up to run over a longer time period.

I did observe the buffer file handle behavior though: the disk buffer files are indeed help open, even after vector has caught up when no new data is coming in. I believe this issue is tracked by https://github.com/timberio/vector/issues/6597 . Vector should be compacting the disk buffer more regularly to free up space and drop unused file handles. I demonstrated this with the following config:

```toml
data_dir = "/tmp/vector"

[sources.in]
type = "socket"
mode = "tcp"
address = "0.0.0.0:12345"

[transforms.remap]
type = "remap"
inputs = ["in"]
source = """
. = parse_json!(.message)
"""

[sinks.out]
type = "http"
inputs = ["remap"]
uri = "http://localhost:8080"
encoding.codec = "ndjson"
buffer.type = "disk"
buffer.max_size = 10737418240
buffer.when_full = "block"

[sources.metrics]
type = "internal_metrics"

[sources.host_metrics]
type = "host_metrics"
scrape_interval_secs = 2

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["host_metrics", "metrics"]
address = "0.0.0.0:9876"
```

I ran `flog -l -f json |  nc 127.0.0.1 12345`, let the buffer fill up to ~500 MB, and then stopped it. I then ran the [`http_test_server`](https://github.com/timberio/http_test_server) to receive the traffic. Vector sent all of the buffered data and then idled.

I then observed that the buffer directory still had data in it:

```
du -h -d 1 /tmp/vector
279M    /tmp/vector/out_buffer
279M    /tmp/vector
```

This number did decrease over time, but has been around there for the last ~5 minutes.

Vector still has ~80 open file handles to the ldb files:

```
lsof -p 32407 | grep 'ldb' | wc -l
79
```

@alpi-ua could you provide some more details about what you are seeing? Are those screenshots when vector is idling?

**Comment by jszwedko** [comment]
@ryn9 could you provide the volume of data that was in the buffer directory as you saw the increasing CPU usage? I'm wondering if it is related to lack of compaction issue I described above.

**Comment by ryn9** [comment]
current data in the directory is only about 30MB

**Comment by alpi-ua** [comment]
> @alpi-ua could you provide some more details about what you are seeing? Are those screenshots when vector is idling?

Yes, vector was idling when screenshots were taken.
Need to say - I can't understand why vector is using buffers at all, while having a stable connection to the server. The idea of buffers is store data when there is no connection at all, is it ?

**Comment by jszwedko** [comment]
@alpi-ua thanks for the additional detail. I'm surprised you would see that many ldb files with a disk buffer of ~12 MB. I saw roughly 100 with 500 MB of buffered data.

Regarding use of the buffer, all data flows through the buffers regardless of whether there is downstream pressure or not. This allows us to provide better guarantees that data won't be lost for the `disk` buffer.  That is, the event will be written to disk before sent to the sink to be sent to the destination so that, if vector crashed, you wouldn't lose any data that was in the sink, but not yet sent. I can see why you would have had that expectation though. I think we could consider allowing that to be configured so that events only get buffered if the sink cannot take any more of them at a given time; just with the trade-off of data guarantees.

@alpi-ua @ryn9 Regarding your specific configs, @binarylogic had the prescient observation that your sources are both types of sources that are able to checkpoint (the `file` and `journald` sources), thus having any buffer in-front of the sink is not really buying you much. It would probably be better to not use a buffer and just let backpressure from the sink flow back to the sources. That is, if the sink can't keep up, the source will just stop, or slow down, reading data. That is, unless you are concerned about the file or journald data getting expunged for some reason before vector is able to read it.

**Comment by jszwedko** [comment]
@ktff do you mind taking a look at this one since you've been working with the buffers? I wasn't able to reproduce it, but maybe you have some insight.

**Comment by ryn9** [comment]
@jszwedko - while my source (file) is able to be checkpointed - my files rotate out quickly and I was depending on vector to lower the possibility of losing out on data. For now - I have recently switched to a memory based buffer and cpu usage appears to be much improved. That said - this is not a viable for us for long term usage.

**Comment by jszwedko** [comment]
So I can verify that if the downstream cannot keep up and vector ends up buffering to disk, it does seem to use a lot of CPU and the memory use seems to increase unbounded over time. I'm wondering if that is similar to what you are seeing @ryn9 and @alpi-ua ? Do you see unbounded memory growth?

I was able to reproduce this via:

Config:

```toml
data_dir = "/var/lib/vector/"

[api]
enabled = true

# Metrics
[sources.internal]
type = "internal_metrics"

[sinks.prometheus]
type = "prometheus_exporter"
inputs = ["internal"]
address = "0.0.0.0:9598"

# CDN
[sources.http_in]
type = "socket"
mode = "tcp"
address = "0.0.0.0:18080"

[sinks.http_out]
type = "http"
inputs = ["http_in"]
encoding.codec = "ndjson"
uri = "http://127.0.0.1:8080"
# Batch
batch.max_bytes = 1048576 # 10 MB
batch.timeout_secs = 5

# Buffer
buffer.max_size = 53687091200 # 50 GB
buffer.type = "disk"
buffer.when_full = "block"
```

I then started up the [`http_test_server`](https://github.com/timberio/http_test_server) via:

```
./http_test_server -m 1000ms -b HARD  -c 5 -d 1s -q 5
```

That is, request latencies of 1000ms and a hard rate limit of 5 requests / second.

Then I ran `flog` with:

```
flog -l -f json |  nc 127.0.0.1 18080
```

Because flog will send faster than the http test server will accept data, vector starts to buffer the events to disk. This is normal. However, as it does that, I see every increasing memory usage that seems to be relative to the size of the buffer. When the buffer was 8.7 GB, Vector was using roughly 10 GB according to `top`. `ps` reported:

```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
CORP\je+ 18680  113 43.0 18522604 14013644 pts/5 Sl+ 21:33   1:01 vector-0.13.1 --config /tmp/http.toml
```

This behavior is even more dramatic if you stop vector when it has a lot of data buffered to disk (around 10GB in my example). `flog` will due since `nc` dies, so nothing will be sending. Then start `vector` back up and observe that it is slow to start and starts consuming a lot of memory before it even starts trying to send it downstream again. This says to me like it is reading the entire buffer into memory.

**Comment by alpi-ua** [comment]
> That is, the event will be written to disk before sent to the sink to be sent to the destination so that, if vector crashed, you wouldn't lose any data that was in the sink, but not yet sent. 
Can you please clarify, vector is writing to the buffer already transformed data, or transformation is happening after buffering?

**Comment by alpi-ua** [comment]
> So I can verify that if the downstream cannot keep up and vector ends up buffering to disk, it does seem to use a lot of CPU and the memory use seems to increase unbounded over time. I'm wondering if that is similar to what you are seeing @ryn9 and @alpi-ua ? Do you see unbounded memory growth?
> 
> I was able to reproduce this via:
> 
> Config:
> 
> ```toml
> data_dir = "/var/lib/vector/"
> 
> [api]
> enabled = true
> 
> # Metrics
> [sources.internal]
> type = "internal_metrics"
> 
> [sinks.prometheus]
> type = "prometheus_exporter"
> inputs = ["internal"]
> address = "0.0.0.0:9598"
> 
> # CDN
> [sources.http_in]
> type = "socket"
> mode = "tcp"
> address = "0.0.0.0:18080"
> 
> [sinks.http_out]
> type = "http"
> inputs = ["http_in"]
> encoding.codec = "ndjson"
> uri = "http://127.0.0.1:8080"
> # Batch
> batch.max_bytes = 1048576 # 10 MB
> batch.timeout_secs = 5
> 
> # Buffer
> buffer.max_size = 53687091200 # 50 GB
> buffer.type = "disk"
> buffer.when_full = "block"
> ```
> 
> I then started up the [`http_test_server`](https://github.com/timberio/http_test_server) via:
> 
> ```
> ./http_test_server -m 1000ms -b HARD  -c 5 -d 1s -q 5
> ```
> 
> That is, request latencies of 1000ms and a hard rate limit of 5 requests / second.
> 
> Then I ran `flog` with:
> 
> ```
> flog -l -f json |  nc 127.0.0.1 18080
> ```
> 
> Because flog will send faster than the http test server will accept data, vector starts to buffer the events to disk. This is normal. However, as it does that, I see every increasing memory usage that seems to be relative to the size of the buffer. When the buffer was 8.7 GB, Vector was using roughly 10 GB according to `top`. `ps` reported:
> 
> ```
> USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
> CORP\je+ 18680  113 43.0 18522604 14013644 pts/5 Sl+ 21:33   1:01 vector-0.13.1 --config /tmp/http.toml
> ```
> 
> This behavior is even more dramatic if you stop vector when it has a lot of data buffered to disk (around 10GB in my example). `flog` will due since `nc` dies, so nothing will be sending. Then start `vector` back up and observe that it is slow to start and starts consuming a lot of memory before it even starts trying to send it downstream again. This says to me like it is reading the entire buffer into memory.
> 
> I tested this with vector 0.13.1 and 0.12.2.

Yes, exactly. 
I see that vector sometimes is using a lot of memory and go back to normal only after restart.

![vector_memory_leak_top](https://user-images.githubusercontent.com/14225967/118446370-3c046e00-b6f8-11eb-8586-ab5476714698.png)
![vector_memory_leak](https://user-images.githubusercontent.com/14225967/118446372-3c9d0480-b6f8-11eb-9f54-a26c60a3241f.png)

**Comment by ktff** [comment]
>  vector is using buffer before the filtering regexp

@alpi-ua  it does't, only sinks can have a buffer, and the increase in memory usage is because of them. #7264 will decrease memory usage for cases when the sink is able to keep up with incoming events. And regarding 

> But there were a lot of them, about 10000 when I noticed a problem.

I'm investigating fragmentation lead on that.

Also, have you seen Vector's memory usage ever deplete available RAM causing the use of swap memory?

**Comment by alpi-ua** [comment]
> Also, have you seen Vector's memory usage ever deplete available RAM causing the use of swap memory?

Yes, up to 600Mb I think

**Comment by ktff** [comment]
@alpi-ua was that with disk buffer or memory buffer? maybe both?

**Comment by alpi-ua** [comment]
You mean Vector buffer? With both disabled as it shown in config 

```
[sinks.app_prod]                                                                                                                                                                                                                                                                          
  type = "vector"                                                                                                                                                                                                                                                                           
  inputs = ["app_filter","app_std"]                                                                                                                                                                                                                                                     
  address = "server:5001"                                                                                                                                                                                                                                                    
  healthcheck = true                                                                                                                                                                                                                                                                        
#  buffer.max_size = 10100000                                                                                                                                                                                                                                                               
#  buffer.type = "disk"                                                                                                                                                                                                                                                                     
#  buffer.when_full = "block"                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                            
[sinks.system]                                                                                                                                                                                                                                                                              
  type = "vector"                                                                                                                                                                                                                                                                           
  inputs = ["journal"]                                                                                                                                                                                                                                                                      
  address = "server:5000"                                                                                                                                                                                                                                                    
  healthcheck = true                                                                                                                                                                                                                                                                        
#  buffer.max_size = 10100000                                                                                                                                                                                                                                                               
#  buffer.type = "disk"                                                                                                                                                                                                                                                                     
#  buffer.when_full = "block"
```

**Comment by ktff** [comment]
@alpi-ua then that's an additional bug, if you are not using those buffers the memory usage shouldn't be so high. 

Could you open a new issue for that, I could compile what you've provided in here and open it but you'll know better.

**Comment by ktff** [comment]
> I still can't understand how Vector can use a large buffer for memory trying to send something thru the sink, if sink has nothing to send, almost everything is getting filtered before.

@alpi-ua yep, it's a bug in both cases, using and not using buffers.

**Comment by alpi-ua** [comment]
> @alpi-ua then that's an additional bug, if you are not using those buffers the memory usage shouldn't be so high.
> 
> Could you open a new issue for that, I could compile what you've provided in here and open it but you'll know better.

https://github.com/timberio/vector/issues/7525

**Comment by chanr32** [comment]
We are also experiencing this issue when trying to use buffered mode. Would be awesome to see a fix for it when possible.

**Comment by ktff** [comment]
@chanr32 there are multiple fixes on the way, some are already merged.

**Comment by ktff** [comment]
@chanr32 @alpi-ua @ryn9 fixes for disk buffers have landed on master and are currently on nightly, while `v0.14` should be released soon. Could you all check it out? 

Also @alpi-ua @ryn9 the underlying issues with CPU and memory consumption should be fixed, with a caveat that the underlying disk isn't too slow like AWS gp2 volume, but it's possible that there are more causes of it that were just hiding behind the fixed issues. 

Also noting that regarding memory consumption, disk buffer will also be held in memory, at least a part of it, for performance reasons. So using 10GB disk buffer can normally consume 11~12GB or RAM when the buffer is full.

**Comment by ktff** [comment]
This is fixed in `v0.14` but we can reopen it if it turns out that there are more causes.
